src/core/midiPorts.cpp

```diff
@@ -0,0 +1,416 @@
+
+		// This port already exists in the map
+		if (outPorts_[port]->isPortOpen()) {
+		
+			// And this port is indeed open, nothing to do.
+			return 2;
+		}
+		// However if it is in the map but not open, something is wrong
+		// Oh well, we'll se what happens next.
+	}
+
+	// No entry in map, so we shall create one
+	else {
+		try {
+			outPorts_[port] = new RtMidiOut((RtMidi::Api) api_,
+						"Giada Output to " + port);
+		}
+		catch (RtMidiError& error) {
+			u::log::print("[MP::openOutPort] RtMidiError: %s\n",
+						error.getMessage().c_str());
+			return -2;
+		}
+	}
+	
+	// Let's open this port
+	try {
+		outPorts_[port]->openPort(index, port);
+		u::log::print("[MP::openOutPort] MIDI port \"%s\" open\n", port);
+		return 1;
+	}
+	catch (RtMidiError& error) {
+		u::log::print("[MP::openOutPort] unable to open port %s: %s\n",
+					port, error.getMessage().c_str());
+		return -3;
+	}
+
+}
+
+/* -------------------------------------------------------------------------- */
+
+int openInPort(std::string port){
+
+	// First let's see if port name is even valid
+	int index = getInDeviceIndex(port);
+	if (index < 0) {
+		// No such port is available, exiting
+		return -1;
+	}
+
+	// Let's check if this port is in the map, possibly already open //
+	if (inPorts_.count(port) > 0) {
+
+		// This port already exists in the map
+		if (inPorts_[port]->isPortOpen()) {
+		
+			// And this port is indeed open, nothing to do.
+			return 2;
+		}
+		// However if it is in the map but not open, something is wrong
+		// Oh well, we'll se what happens next.
+	}
+
+	// No entry in map, so we shall create one
+	else {
+		try {
+			inPorts_[port] = new RtMidiIn((RtMidi::Api) api_,
+						"Giada Input to " + port);
+		}
+		catch (RtMidiError& error) {
+			u::log::print("[MP::openInPort] RtMidiError: %s\n",
+						error.getMessage().c_str());
+			return -2;
+		}
+	}
+	
+	// Let's open this port
+	try {
+		// Create a persistent string with a port name, so callback
+		// function knows which port did the data come from. 
+		// TODO: How do we dispose of this thing when the port is closed?
+		// A unique_ptr would be nice but I think it's not applicable here.
+		std::string* p = new std::string(port);
+
+		inPorts_[port]->setCallback(&callback_, p);
+		inPorts_[port]->openPort(index, port);
+		inPorts_[port]->ignoreTypes(false, false, false);
+		u::log::print("[MP::openInPort] MIDI port \"%s\" open\n", port);
+		return 1;
+	}
+	catch (RtMidiError& error) {
+		u::log::print("[MP::openInPort] unable to open port %s: %s\n",
+					port, error.getMessage().c_str());
+		return -3;
+	}
+
+}
+
+/* -------------------------------------------------------------------------- */
+
+int closeOutPort(std::string port){
+
+	if (port == ""){
+
+		// Create a list of open ports and destroy them all!
+		std::map<std::string, RtMidiOut*>::iterator it;
+		std::vector<std::string> ports_to_destroy;
+
+		for (it = outPorts_.begin(); it != outPorts_.end(); ++it){
```

Comment:
In C++17 you can iterate over maps with structured bindings: https://stackoverflow.com/a/26282004/3296421